compiler/src/dotty/tools/dotc/ast/Trees.scala

```diff
@@ -754,16 +753,24 @@ object Trees {
     def isClassDef: Boolean = rhs.isInstanceOf[Template[_]]
   }
 
-  /** extends parents { self => body } */
-  case class Template[-T >: Untyped] private[ast] (constr: DefDef[T], parents: List[Tree[T]], self: ValDef[T], private var preBody: LazyTreeList)
+  /** extends parents { self => body }
+   *  @param parentsOrDerived   A list of parents followed by a list of derived classes,
+   *                            if this is of class untpd.DerivingTemplate.
+   *                            Typed templates only have parents.
+   */
+  case class Template[-T >: Untyped] private[ast] (constr: DefDef[T], parentsOrDerived: List[Tree[T]], self: ValDef[T], private var preBody: LazyTreeList)
     extends DefTree[T] with WithLazyField[List[Tree[T]]] {
     type ThisTree[-T >: Untyped] = Template[T]
     def unforcedBody: LazyTreeList = unforced
     def unforced: LazyTreeList = preBody
     protected def force(x: AnyRef): Unit = preBody = x
     def body(implicit ctx: Context): List[Tree[T]] = forceIfLazy
+
+    def parents: List[Tree[T]] = parentsOrDerived // overridden by DerivingTemplate
+    def derived: List[untpd.Tree] = Nil           // overridden by DerivingTemplate
```

Comment:
`derived` only exist as untyped trees. Typed trees contain `derived$...` members instead. It would be pointless to keep typed `derived` trees around, as the info in them is never relevant.